### DISCUSSION TITLE
Frigate web front end authentication

### DISCUSSION THREAD

**Discussion by blacknell** [blakeblackshear/frigate#15414] [Ask A Question] [answered]
I'd like to enable Fail2ban to run against the Frigate front end logs so we can pick up rogue login attempts.

My goal is to document a 'how to' so other users can enable if required. I've noticed a few requests in the past.

My initial question relates to the front end platform/framework. Could someone let me know what this is built on. From what I can see in the Nginx logs it's returning a 400 error on failed login and I couldn't easily see how this was generated.

**Comment by NickM-27** [comment]
The 400 is just sent when the login API gets incorrect credentials

**Comment by blacknell** [comment]
That’s what I thought. Shouldn’t it be a 403RegardsPaulOn 9 Dec 2024, at 12:59, Nicolas Mowen ***@***.***> wrote:﻿
The 400 is just sent when the login API gets incorrect credentials

**Reply by NickM-27** [reply]
there may be a reason it was 400 specifically, CC @blakeblackshear

**Comment by blakeblackshear** [comment]
Is there a reason the built-in protection doesn't suit your needs? https://docs.frigate.video/configuration/authentication#login-failure-rate-limiting

This already blocks repeated login attempts from the same ip address.

**Reply by blacknell** [reply]
I think the protection is offered good.

But my preference is to use Fail2Ban to block all access from a host that tries brute forcing. As is pervasive out there, a lot of attacks happen over time so I'd rather just drop that IP completely at the firewall.

I report via Abuse IPDB as well which contributes to the greater good.

I'd like to offer this into the docs for those that want it, if you're in agreement. I've got the filters working here.

**Reply by blakeblackshear** [reply]
We could add a message to the logs that's compatible with fail2ban with the ip address which might help. 

On your other questions, I typically see 400 as the response when an incorrect password is used. 401s are used to indicate that authentication is missing but needed. A 403 would be sent if the IP was being blocked or rate limited by the API.

My understanding is that this is the correct use of the http responses. 403 would not be correct when attempting to login with the wrong password. A 403 would indicate the client is forbidden from calling the login endpoint at all.

**Reply by blacknell** [reply]
I don't think the log output needs changing. Since this is being reverse proxied by (in my case) apache the ip address is placed in it's log file along with the output from frigate. So fail2ban has everything it needs. I personally don't want frigate directly facing the internet - I'd rather have my apache reverse proxy running on a bastion brokering such requests.

eg.
```logs
**REDACTED**:443 192.168.1.13 - - [08/Dec/2024:10:37:44 +0000] "POST /api/login HTTP/1.1" 400 530 "https://**REDACTED*/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
```

I agree 403 is probably wrong but 401 is the correct code because if the request included authorisation credentials and they're bad then this is the meaning of the response. If no credentials supplied then 401 response should also include further auth challenges and apache doesn't log as an error. 403 means the resource is denied for that authorised user but it also means failed login when the request included the WWW-Authenticate header field challenge following the previous 401. 429 is for rate limiting.

All in all it's not entirely obvious.

Looking at other fail2ban filters they seem to mostly ignore the actual code and look for regex in the output string eg. Authorization of user paul failed.

My suggestion is to leave it as 400 but commit to leaving the request as `/api/login`

My regex is looking for `^.+?:\d+ <HOST> -.* "(GET|POST|HEAD) /api/login *HTTP.*" 400 .*$` and seems to work perfectly.

I was more worried about the log output changing in the future

**Reply by blakeblackshear** [reply, marked as answer]
I don't see any reason the path with change in the future.

After reading [this](https://auth0.com/blog/forbidden-unauthorized-http-status-codes/), I agree that it makes sense to respond with a 401 when invalid credentials are submitted.

**Reply by blacknell** [reply]
Ok, great. 

If you're in agreement I'll submit a documentation PR on how to support Fail2ban - though it'll be limited to Apache reverse proxy as I don't have any experience and can't test Nginx.

**Reply by blakeblackshear** [reply]
That's fine. Probably best under advanced topics.

**Reply by blakeblackshear** [reply]
You could also post it as a show and tell discussion

**Reply by blacknell** [reply]
do you think this can make it to the next beta?

**Reply by NickM-27** [reply]
What are you referring to?

**Reply by blacknell** [reply]
@blakeblackshear suggested the return code after failed login would be changed to 401

**Reply by NickM-27** [reply]
that has already been implemented for beta 3

**Reply by hawkeye217** [reply]
https://github.com/blakeblackshear/frigate/pull/15432

This was already merged and is in beta 3.
